include/ignition/gazebo/detail/BaseView.hh

```diff
@@ -0,0 +1,199 @@
+
+  /// \brief Remove an entity from the view.
+  /// \param[in] _entity The entity to remove.
+  /// \return True if the entity was removed, false if the entity did not
+  /// exist in the view.
+  public: virtual bool RemoveEntity(const Entity _entity) = 0;
+
+  /// \brief Add the entity to the list of entities to be removed
+  /// \param[in] _entity The entity to add.
+  /// \return True if the entity was added to the list, false if the entity
+  /// was not associated with the view.
+  public: bool AddEntityToRemoved(const Entity _entity);
+
+  /// \brief Clear the list of new entities
+  public: void ClearNewEntities();
+
+  /// \brief Get the set of component types that this view requires.
+  /// \return The set of component types.
+  public: const std::set<ComponentTypeId> &ComponentTypes() const;
+
+  /// \brief Clear all data from the view and reset it to its original, empty
+  /// state.
+  public: virtual void Reset() = 0;
+
+  /// \brief Get all of the entities in the view
+  /// \return The entities in the view
+  public: const std::set<Entity> &Entities() const;
+
+  /// \brief Get all of the new entities in the view
+  /// \return The new entities in the view
+  public: const std::set<Entity> &NewEntities() const;
+
+  /// \brief Get all of the entities to be removed from the view
+  /// \return The entities to be removed from the view
+  public: const std::set<Entity> &ToRemoveEntities() const;
+
+  /// \brief Get all of the entities to be added to the view
+  /// \return The entities to be added to the view
+  public: const std::unordered_map<Entity, bool> &ToAddEntities() const;
+
+  /// \brief Clear all of the entities that are marked as to be added to the
+  /// view
+  public: void ClearToAddEntities();
+
+  /// \brief Destructor
+  public: virtual ~BaseView()
+  {
+  }
+
+  // TODO(adlarkin) make this a std::unordered_set for better performance.
+  // We need to make sure nothing else depends on the ordered preserved by
+  // std::set first
+  /// \brief All the entities that belong to this view.
+  protected: std::set<Entity> entities;
+
+  // TODO(adlarkin) make this a std::unordered_set for better performance.
+  // We need to make sure nothing else depends on the ordered preserved by
+  // std::set first
+  /// \brief List of newly created entities
+  protected: std::set<Entity> newEntities;
+
+  // TODO(adlarkin) make this a std::unordered_set for better performance.
+  // We need to make sure nothing else depends on the ordered preserved by
+  // std::set first
+  /// \brief List of entities about to be removed
+  protected: std::set<Entity> toRemoveEntities;
+
+  /// \brief List of entities to be added to the view. The value of the map
+  /// indicates whether the entity is new to the entity component manager or not
+  protected: std::unordered_map<Entity, bool> toAddEntities;
+
+  /// \brief The component types in the view
+  protected: std::set<ComponentTypeId> componentTypes;
```

Comment:
This is used with the `EntityMatches` method (notice how the `_types` parameter is a `std::set`): https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/include/ignition/gazebo/EntityComponentManager.hh#L143-L148

Here's an example use case: https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/include/ignition/gazebo/detail/EntityComponentManager.hh#L504-L506

I could change this to be `unordered`, but that would also mean another `EntityMatches` method with a `std::unordered_set` parameter type needs to be created, and the existing `EntityMatches` method that takes a `std::set` should be deprecated. If you want me to make this change, I've got no problem with it, but I figured that it may not be worth going through another deprecation because I don't think the performance gains of using a `std::unoredered_set` for `componentTypes` will be very noticeable - `componentTypes` is populated when the view is created, and is never modified after that: https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/include/ignition/gazebo/detail/View.hh#L49-L53

So, considering the fact that 1) we only need to insert elements into `componentTypes` once, 2) the elements are a primitive type (`uint64_t`), which reduces complexity for insertion comparison, and 3) it's unlikely that the number of elements in `componentTypes` will be very large (most views consist of less than 10 types), I wasn't sure if adding a new `EntityMatches` method (and deprecating the old one) was worth it for _what I believe to be_ a very minor optimization (it'd be a `O(1)` operation instead of `O(logn)` for a few insertions that only happens once).

What do you think? Should we change this to `std::unordered_set` and add a new `EntityMatches` method? Another option would be to not deprecate the old `EntityMatches` method and keep the two methods with different input parameter types, but I don't see the value in keeping two different method signatures here.